text/0000-implied-derive.md

```diff
@@ -0,0 +1,425 @@
+With this RFC, we can get rid of needless mention of super traits,
+and instead write:
+
+```rust
+#[derive(Copy, Ord, Debug, Hash)]
+pub enum Option<T> {
+    None,
+    Some(T),
+}
+```
+
+This definition is significantly more terse and thus more ergonomic when
+rapidly prototyping. As a benefit to improved ergonomics, newtypes become
+more encouraged as you have to explicitly derive fewer traits.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Vocabulary and definitions
+
+- The **cartesian product** `S × T` of the sets `S` and `T`
+  is defined as `S × T = {(s, t) | s ∈ S ∧ t ∈ T}`.
+
+- A **binary relation** `R` on sets `S` and `T` is a subset of `S × T`.
+
+- A **transitive relation** `R ⊆ A × A` is a binary relation such that:
+  `∀ x, y, z ∈ A. (x, y) ∈ R ∧ (y, z) ∈ R => (x, z) ∈ R`.
+  That is: for all elements `x`, `y`, and `z` in the set `A`,
+  if `x` and `y` are related, and `y` and `z` are related,
+  then `x` and `z` must also be related.
+
+- The **transitive closure** `R+` of a transitive relation `R` is defined as:
+  ```
+  R_0       = R
+  R_{i + 1} = R_i ∪ {(s, u) | ∃ t. (s, t) ∈ R_i ∧ (t, u) ∈ R_i}
+  ```
+  From the perspective of an element `x` related to some `y` via `R`,
+  i.e: `x R y`, the transitive closure can be seen as all the `y`s you can
+  reach from `x` in one or more steps.
+
+- **super trait** - For a trait `Copy`, defined as `trait Copy : Clone {}`,
+  the trait `Clone` is a super trait of `Clone`. We also say that if `T: Copy`,
+  for some type `T`, then `T: Clone`, or in other words:
+  `Copy` implies `Clone` (denoted `Copy => Clone`).
+
+## New concepts
+
+- An **implicitly derived** trait is a trait `T` which gets implemented for
+  a type because the type has `#[derive(S)]`, where `T` is in the transitive
+  super-trait closure of `T`.
+
+## Practical implications
+
+Previously, it was not enough to `#[derive(Copy)]` to also implement `Clone`
+for the type being derived. Instead, you had to `#[derive(Copy, Clone)]`.
+With this RFC, you can `#[derive(Copy)]`, and then `#[derive(Clone)]` is implied.
+This change also applies to `Ord` and all other, now or future,
+derivable standard library traits.
+
+In other words, we have that:
+
++ `#[derive(Copy)]` => `#[derive(Clone)]`
++ `#[derive(Ord)]` => `#[derive(Eq, PartialEq, PartialOrd)]`
++ `#[derive(PartialOrd)]` => `#[derive(PartialEq)]`
++ `#[derive(Eq)]` => `#[derive(PartialEq)]`
+
+### Lints
+
++ If you explicitly derive a trait that can be implicitly derived,
+  then a warning named along the lines of `explicitly_derived_super_trait`
+  will be issued recommending that you remove the implictly derived trait.
+  You are encouraged as a user to heed this warning.
+
+### Errors
+
++ `#[derive(Clone, Clone)]` will no longer issue an error about conflicting
+  implementations.
+
+### Custom derive
+
++ This RFC does **not** affect the behavior of `#[derive(..)]` for traits
```

Comment:
This is unfortunate, but does answer the question I was going to ask about how trait resolution comes into play

Reply:
See also the section *"Future work"* wrt. [`#[proc_macro_derive(Sub, implies(Super))]`](https://github.com/Centril/rfcs/blob/rfc/implied-derive/text/0000-implied-derive.md#proc_macro_derivesub-impliessuper).